### PR TITLE
Fix MERGE bug when join condition is false

### LINKED ISSUE #13751: SQL MERGE fails if ON condition is literal FALSE
@przemekak reported that execution of an SQL MERGE statement of this form fails:

```
MERGE INTO target_table t USING source_table s
ON FALSE
WHEN NOT MATCHED THEN INSERT (...) VALUES (...)
```

The failure is due to elimination of the RIGHT JOIN because the join predicate is false.  The fix is to recognize during optimization that the JOIN has been optimized away, and turn the MERGE into an INSERT operation.

### PR DESCRIPTION
**Pull request by djsstarburst**
## Description

The bug showed up as an exception in BeginTableWrite.rewriteModifyTableScan
because it can't find the TableScanNode for the target table.  For MERGE,
if the join condition is not false, Some connectors need to find the
TableScanNode to update the table handle in the MergeTarget.  However,
if the join condition is shown by the optimizer to be false, the
MergeTarget handle need not be updated.

> Is this change a fix, improvement, new feature, refactoring, or other?
It's a bug fix.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
It's a change to the core query engine.

> How would you describe this change to a non-technical end user or system administrator?
Fix a bug that caused SQL MERGE operations to fail if the optimizer could prove the join conditions was false.

## Related issues, pull requests, and links

Fixes #13751

## Documentation

(X) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(X) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 1]
```diff
@@ -4110,6 +4110,47 @@ public void testMergeAllInserts()
...
+        assertUpdate(createTableForWrites(format("CREATE TABLE %s (customer VARCHAR, purchases INT, address VARCHAR)", targetTable)));
+
+        assertUpdate(format("INSERT INTO %s (customer, purchases, address) VALUES ('Aaron', 11, 'Antioch'), ('Bill', 7, 'Buena')", targetTable), 2);
+
+        // Test a literal false
+        assertUpdate(format("MERGE INTO %s t USING ", targetTable) +
+                        "(VALUES ('Carol', 9, 'Centreville')) AS s(customer, purchases, address)" +
+                        " ON (FALSE)" +
+                        "    WHEN NOT MATCHED THEN INSERT (customer, purchases, address) VALUES(s.customer, s.purchases, s.address)",
+                1);
```
Comment (martint) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:4128]: You can use multiline strings for these tests:

```java
 assertUpdate("""
        MERGE INTO %s t USING 
            (VALUES ('Carol', 9, 'Centreville')) AS s(customer, purchases, address)
        ON (FALSE)
        WHEN NOT MATCHED THEN 
            INSERT (customer, purchases, address) VALUES(s.customer, s.purchases, s.address)
    """.formatted(targetTable),
    1);
```

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java [hunk 2]
```diff
@@ -184,7 +184,7 @@ public PlanNode visitMergeWriter(MergeWriterNode mergeNode, RewriteContext<Optio
             MergeTarget mergeTarget = (MergeTarget) getContextTarget(context);
             return new MergeWriterNode(
                     mergeNode.getId(),
-                    rewriteModifyTableScan(mergeNode.getSource(), mergeTarget.getHandle()),
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java:187]: if so, the Merge's TS should not return true from `io.trino.sql.planner.plan.TableScanNode#isUpdateTarget`

**Comment by djsstarburst** [comment]
Removing the `rewriteTableScan` caused the Hive tests to fail because the `HiveTableHandle` in the `MergeTarget` doesn't have a running `AcidTransaction` until `BeginTableWrite.createWriterTarget` calls `ConnectorMetadata.beginMerge`. The merge handle returned by that method has the `AcidTransaction` initialized, and `rewriteTableScan` is used to update the handle in the `TableScanNode`.

Why Hive care if the table handle in the `TableScanNode` got updated with the `AcidTransaction`? Two reasons. First, the code tests in various places that a transaction is running by examining the table handle. Second, the `AcidTransaction` object contains the `transactionId` with is used in a bunch of places in `SemiTransactionalHiveMetastore`.  (It also has the writeId, but that's only accessed by `UpdatablePageSource`, which isn't used in MERGE.)

Perhaps at some point we can change the Hive transaction object for modify operations to be global and not hidden in the table scan, since AIUI we only allow one modify operation at a time. But that's a big, potentially dangerous change.

The simplest alternative is to look for the table scan to modify, but not throw an exception if we don't find it for MERGE. I'm going to try that to verify that it fixes all the tests.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 1 above)
Comment (djsstarburst) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: Changed.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java (hunk 2 above)
Comment (djsstarburst) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java:187]: As you may have gathered from the comment below, I was wrong about needing to find the right `TableScanNode` **if it exists**.

**Comment by djsstarburst** [comment]
Merged as part of PR #14636.

Closing this PR.

**Pull request closed without merging**